### ISSUE TITLE
New IR "unknown" type?

### ISSUE THREAD

**Issue by Laaas** [raptorjit/raptorjit#72]
I think an IR type, that represents an unknown type, would be beneficial in many cases.
It could be valid for only a subset of all IR operations, e.g.:
NE
EQ
All the LOAD and STORE operations

This would be very useful in code where the type does not matter, e.g.:
```lua
for i = 1, #some_table do
  if some_table[i] == some_other_table then
    break
  end
end
```
This is currently compiled to the following IR with LuaJIT 2.1 alpha 3 (I have no idea how to use jit.dump with RaptorJIT), if it's filled with numbers:
```
....        SNAP   #0   [ ---- ]
0001 >  int SLOAD  #4    CRI
0002 >  int LE     0001  +2147483646
0003    int SLOAD  #3    CI
0004 >  tab SLOAD  #2    T
0005    int FLOAD  0004  tab.asize
0006 >  p32 ABC    0005  0001
0007    p32 FLOAD  0004  tab.array
0008    p32 AREF   0007  0003
0009 >  num ALOAD  0008
0010  + int ADD    0003  +1  
....        SNAP   #1   [ ---- ---- ---- ]
0011 >  int LE     0010  0001
....        SNAP   #2   [ ---- ---- ---- 0010 0001 ---- 0010 ]
0012 ------ LOOP ------------
0013    p32 AREF   0007  0010
0014 >  num ALOAD  0013
0015  + int ADD    0010  +1  
....        SNAP   #3   [ ---- ---- ---- ]
0016 >  int LE     0015  0001
0017    int PHI    0010  0015
```
The problem here, is that 0014 typechecks 0013, and asserts that it's a number.
This isn't such a big of a problem here, since the typecheck is inexpensive, but if the table contains a bunch of different types, the trace will abort, no?
What it should really compile the inner loop as is:
```
0013    p32 AREF   0007  0010
0014    unk ALOAD  0013
0015 >  unk NE     0014 <some_other_table>
0016  + int ADD    0010  +1
....        SNAP   #3   [ ---- ---- ---- ]
0017 >  int LE     0015  0001
0018    int PHI    0010  0015
```

**Comment by lukego** [comment]
This idea sounds highly plausible to me! Thanks for sharing it. I can imaging that some code will JIT poorly today and would be helped by this change. Maybe it would be an optimization pass similar to `narrow` that replaces IR types with `unk` where compatible?

I made a first attempt to define a benchmark to exercise a bad case in the JIT today, using LuaJIT 2.1. I failed because, damn, the JIT has some impressive tricks :). Let me share how the JIT nailed this first benchmark, then I will try harder to find a bad case.

The example is simple. First create a heterogeneous array with all elements of different types:

```lua
local array1 = {1, 'one', {1}, require("ffi").typeof('int'), function() end}
```

Now write a function that will manipulate the array without caring about the types of the elements:

```
local function reverse (array)
  local n = #array
  for i = 1, math.floor(n/2) do
    array[i], array[n-i] = array[n-i], array[i]
  end
end
```

Then call the function in a loop. I expected an explosion of side-traces because separate code has to be generated for each element type, due to the lack of the `unk` wildcard type. However, the JIT generated only two traces. The first one unrolled the loop over the array elements and specialized the type of each load separately. Here are the `ALOAD` instructions:

```lua
cdt ALOAD  0060
num ALOAD  0032
tab ALOAD  0068
str ALOAD  0043
```

Pretty neat, eh? However: on the next iteration the types are different and so this trace aborts. The JIT then simply created a second trace with the other order for the types, which then links back to the first trace.

So the `reverse` function compiles to two traces, one that reverses the array and another that un-reverses it, and execution is interleaved from one to the next.

Trickier than I had expected :).

Let's see what happens if we make it a rotate instead of a reverse though...

**Comment by Laaas** [comment]
I thought about it for a while and I think I have a way to implement it.
Also: why do the STORE operations even *have* types? They're completely type-agnostic, no?

I think a good way to implement it would be to have a whole bit for declaring a type as unknown.
In lj_ir.h I can see that the guard and phi operand instructions are implemented this way, but I don't know if there's place for yet another bit.

Basically, all instructions would default to having this bit set, and then each instruction that uses this instruction would clear this bit, if it needs a definite type. Then at the code-gen stage, would the type-checks be omitted.

**Comment by Laaas** [comment]
Actually, maybe it doesn't need a new bit at all, maybe the GUARD bit can be reused?
This way, stray LOADs with unknown types can be completely eliminated!

This does complicate the implementation though, since if each instruction sets the GUARD bit for each used instruction, the number of type checks would sky-rocket. So there needs to be a special case for LOADs, and an easy an quick way to distinguish them from the other instructions.

**Comment by Laaas** [comment]
Another problem though, is the recording stage. I don't know much about how this stage records the type, and all I can do is guess. How would one stop the trace from aborting with any of these three implementations? (Distinct unknown type, unknown bit, reusing GUARD bit)

**Comment by Laaas** [comment]
I realised something while thinking about #75 
Let's say we have the following IR:

```
0001    num SLOAD  #1    R
0002    num SLOAD  #2    R
0003    num SLOAD  #3    R
0004    num SLOAD  #4    R
0005    num EQ     0003  0004
; no type-dependent instructions are using 0001-0004
0006 >  num SEL    0005  0001  0002
0007    num ADD    0006  +1
; 0006 needs to have GUARD bit set, better than having 4 though!
; We only set the GUARD bit for loads, stores, and SEL.
```
Here, there's no problem. We just have to set the GUARD bit for SEL too. But if the sources of SEL are guaranteed to be their respective types, then there will be duplicate type checks.

**Comment by Laaas** [comment]
Do you guys know what IRSLOAD_TYPECHECK does?
Does it do what it sounds like it's doing, e.g. controlling type checking, thus rendering not having a GUARD on it obsolete? Or is there already no guard on non-type-checked stack loads?

**Comment by Laaas** [comment]
Yep it doesn't do type checking when that bit is not present.

**Comment by igelhaus** [comment]
I may be late to the party, but this case hurts us badly, at least in some projects.

Our typical project is a data-heavy application, with lots of decisioning and relatively small amount of actual computations. So whenever we have to perform an operation like field-by-field table copying (e.g. we have some internal data structures, and we want to create a new object for sending it to a partner via a HTTP request), we face with trace explosions in this particular region of code.
